File: PathSumII.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        vector<vector<int> > res;
        vector<int> cur;
        
        pathSumHelper(root, res,cur, sum);
        
        return res;
    }
private:
    void pathSumHelper(TreeNode * node, vector<vector<int> > &res, vector<int> &cur, int sum) {
        if (!node)
            return;
        if (!node->left && !node->right) {
            sum -= node->val;
            if(!sum) {
                cur.push_back(node->val);
                res.push_back(cur);
                cur.pop_back();
            }
            return;
            
        }
        
        if (node->left) {
            
            sum -= node->val;
            if (true) {
                cur.push_back(node->val);
                pathSumHelper(node->left, res, cur, sum);
                cur.pop_back();
            }
            
            sum += node->val;
        }
        
        if (node->right) {
            
            sum -= node->val;
            if (true) {
                cur.push_back(node->val);
                pathSumHelper(node->right, res, cur, sum);
                cur.pop_back();
            }
            sum += node->val;
        }
        
        
        
    }
};

File: jiuzhang/path-sum-ii/path-sum-ii.java
/**
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班，
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
 */

public class Solution {
    public ArrayList> pathSum(TreeNode root, int sum) {
        ArrayList> rst = new ArrayList<ArrayList<Integer>>();
        ArrayList solution = new ArrayList<Integer>();

        findSum(rst, solution, root, sum);
        return rst;
    }

    private void findSum(ArrayList> result, ArrayList<Integer> solution, TreeNode root, int sum){
        if (root == null) {
            return;
        }

        sum -= root.val;

        if (root.left == null && root.right == null) {
            if (sum == 0){
                solution.add(root.val);
                result.add(new ArrayList(solution));
                solution.remove(solution.size()-1);
            }
            return;
        }

        solution.add(root.val);
        findSum(result, solution, root.left, sum);
        findSum(result, solution, root.right, sum);
        solution.remove(solution.size()-1);
    }
}
